### ISSUE TITLE
Support for multiple partitions

### ISSUE THREAD

**Issue by terra-yuri** [raspberrypi/rpi-sb-provisioner#78]
Hi,

I have a case of 3 partitions (boot/system/data) instead of 2 (boot/system) in the OS image.

Will the provisioner support it as-is, and LUKS encrypt both the system+data partitions? if not, what should be changed to add this functionality? 

Thank you!

**Comment by tdewey-rpi** [comment]
This isn't explicitly supported today - but I'm planning to add a storage-description document that would give structure to this idea and introduce native support.

If you cannot wait for that, I'd suggest modifying the partition mounts and Fastboot command set:

Partition mounts, which I'd suggest you expand to include all partitions you want to flash. Immediately after this site is a good place to perform in-flight modifications - another future change I'm expecting to introduce.
https://github.com/raspberrypi/rpi-sb-provisioner/blob/b8dfe1550e06f85aabddfb3ed59c4a81cd0e3ef2/service/rpi-sb-provisioner.sh#L642

Then the fastboot partitioning options. Here, I'd suggest using the same guidance available on StackOverflow: https://unix.stackexchange.com/questions/482590/how-can-i-create-multiple-partitions-in-a-luks-container-using-a-gpt-mbr-partiti

And have 'partinit' run a second time inside the LUKS container created by 'cryptinit/cryptopen'.
https://github.com/raspberrypi/rpi-sb-provisioner/blob/b8dfe1550e06f85aabddfb3ed59c4a81cd0e3ef2/service/rpi-sb-provisioner.sh#L825

Finally, you'll need to get your data into the correct location. Do this by modifying the Fastboot command set, and don't forget to update the device-mapper names to reflect your new partitions:

https://github.com/raspberrypi/rpi-sb-provisioner/blob/b8dfe1550e06f85aabddfb3ed59c4a81cd0e3ef2/service/rpi-sb-provisioner.sh#L857

As mentioned, as part of the work we're doing in `rpi-image-gen`, I expect to introduce infrastructure to formalise this process.

**Comment by torntrousers** [comment]
> This isn't explicitly supported today - but I'm planning to add a storage-description document that would give structure to this idea and introduce native support.

might that then enable something like A/B partitions and OTA updates of the OS?

**Comment by tdewey-rpi** [comment]
> > This isn't explicitly supported today - but I'm planning to add a storage-description document that would give structure to this idea and introduce native support.
> 
> might that then enable something like A/B partitions and OTA updates of the OS?

Among other storage topologies, yes. Under the current scheme you can use the apt-based upgrade mechanism and in-place updates as usual - however for high-availability deployments and ones where you need to provide trivial device reset, these new storage layouts might offer performance or flexibility benefits.

Expect more information in this area soon, but also take a look at `rpi-image-gen` to see what we're thinking about.

**Comment by tdewey-rpi** [comment]
Both projects will get modifications:

- rpi-image-gen to create a description of provisioning actions (partitioning, LUKS formatting, etc)
- rpi-sb-provisioner to read this description and expand support to fit the description format.

**Comment by jpiwek** [comment]
> Both projects will get modifications:
> 
>     * rpi-image-gen to create a description of provisioning actions (partitioning, LUKS formatting, etc)
> 
>     * rpi-sb-provisioner to read this description and expand support to fit the description format.
@tdewey-rpi : What is the status on this topic or plans implement that? 

There are two image examples listed in rpi-image-gen here: [Link](https://github.com/raspberrypi/rpi-image-gen/tree/92ee07591b3ea22e41ba5299a7b09efd69e931a3/image)

@tdewey-rpi: can those files be used as input for rpi-sb-provisioner? What is the status on this topic or plans implement that?

**Comment by tdewey-rpi** [comment]
> > Both projects will get modifications:
> > ```
> > * rpi-image-gen to create a description of provisioning actions (partitioning, LUKS formatting, etc)
> > 
> > * rpi-sb-provisioner to read this description and expand support to fit the description format.
> > ```
> 
> There are two image examples listed in rpi-image-gen here: [Link](https://github.com/raspberrypi/rpi-image-gen/tree/92ee07591b3ea22e41ba5299a7b09efd69e931a3/image)

Yes, these will produce multi-partition images today, but sb-provisioner still requires changes to support this flow.

> 
> @tdewey-rpi: can those files be used as input for rpi-sb-provisioner? What is the status on this topic or plans to implement that?

Not currently - but this feature is moving up my priority stack pretty quickly.

**Comment by jpiwek** [comment]
@tdewey-rpi: sounds great. Let me know if you need any help with testing

**Comment by jsullivan-mitre** [comment]
Any updates on this being supported in rpi-sb-provisioner? Would be a big help for a project I'm working on at the moment.

Thank you!

**Comment by tdewey-rpi** [comment]
This is now implemented via the IDP support in the 2.3 release series. Happy provisioning!
